mava/networks/retention.py

```diff
@@ -0,0 +1,323 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Tuple
+
+import flax.linen as nn
+import jax
+import jax.numpy as jnp
+from chex import Array
+from omegaconf import DictConfig
+
+from mava.networks.utils.sable import PositionalEncoding
+
+# General shapes legend:
+# B: batch size
+# N: number of agents
+# S: sequence length
+# C: chunk size - T * N in a chunk
+# T: number of timesteps
+
+
+class SimpleRetention(nn.Module):
+    """Simple retention mechanism for Sable.
+
+    Note:
+        This retention mechanism implementation is based on the following code:
+        https://github.com/Jamie-Stirling/RetNet/blob/main/src/retention.py
+    """
+
+    embed_dim: int
+    head_size: int
+    n_agents: int
+    masked: bool
+    decay_kappa: float
```

Comment:
Nice to have if someone is coming just looking at retention file and retnet paper
```suggestion
    decay_kappa: float  # this is gamma in the original retention implementation
```